### DISCUSSION TITLE
TPU support

### DISCUSSION THREAD

**Discussion by gene-git** [Motion-Project/motionplus#154] [Enhancements]
I am not familiar with this, but it might be quite useful / interesting to support TPUs like this little USB one:

[coral usb tpu](https://coral.ai/products/accelerator)

[lib support](https://coral.ai/docs/accelerator/get-started/#requirements)

**Comment by gene-git** [comment]
This tpu may not be good choice - it's hard to find and may not even be being produced at this time.

That said, I think it worthwhile to look at functionality offered by frigate.  Some of things it has would be nice features for motionplus to consider.  For example:

 -  object recognition (people, vehicles, animals). Object detection can be done on cpu or with help from gpu. 
 -  saving object or motion or both.
 -  home assistant / mqtt
 -  media cleanup
 
You're probably fully aware of the other tools out there but thought I'd mention this one which I happened to come across. 

thank you

**Reply by Mr-Dave** [reply]
Yea...the first item is really the important one (object recognition).  It is somewhere on the Motionplus enhancements list.  Really need to get a good model for the OpenCV DNN secondary detection and make that the default in Motionplus.  (And then after that is working, see whether CPU vs GPU processing needs to be addressed).

**Reply by SebastianArcus** [reply]
In case it helps a bit, I have written a basic Python script to add object detection to Motion. I used Yolo from Ultralytics, as it seems to be one of the more efficient and lightweight models/libraries for object recognition. More details here: https://github.com/xj25vm/MotionSpot/

**Reply by Mr-Dave** [reply]
@xj25vm Do you have a trained model that you've tested that I could use as a sample?  (In addition to just seeing using a sample, I'm also interested in seeing scripts/models like these included in the distribution of Motionplus so it can be the "default" for all users.)

**Reply by SebastianArcus** [reply]
I have included further details at the page above. I have used the pre-trained models provided by Ultralytics  - mainly yolov8n (nano) and yolov8s (small). Anything larger is more accurate, but too heavy on hardware requirements for my setup. The models can be downloaded from this page, further down in the "Performance" section: https://docs.ultralytics.com/models/yolov8/ Alternatively, they also download automatically when attempting to use the Yolo Ultralytics Python module. I believe they are trained against the Coco data set - and they work quite well for me. The list of object classes they can detect is here: https://stackoverflow.com/questions/77477793/class-ids-and-their-relevant-class-names-for-yolov8-model - or by running the command suggested at this page in Python against the Yolo module.

I am happy for people to use any parts or ideas from my script above - anything which would help add object detection in Motion / MotionPlus would be welcomed progress

**Reply by gabbymouth** [reply]
@xj25vm You might want to look into https://github.com/xnorpx/blue-onyx as well. There is a Linux version and instead of loading the model file you could issue a python call similar to this: 

`response = requests.post("http://localhost:32168/v1/vision/detection",files={"image":image_data},data={"min_confidence":0.42}).json()`

I have wrote personal use python scripts using Ultralytics models up to yolo11 and blue-onyx that work with motionplus.

**Reply by SebastianArcus** [reply]
Thank you for the suggestion. I suppose the main advantage would be that Blue-onyx can use the GPU or a TPU for object detection - to increase performance. On the downside, it adds a lot of complexity with it needing to run in Docker. Also, while testing MotionSpot I realised that any object detection solution used for cctv would need to do object tracking as well, not only object detection - as, for example, simply detecting the presence of a car in a cctv setup is not of much practical value. You would want to to know when the car is moving. For detecting people or animals, simple object detection is enough though. It doesn't seem Blue-onyx offers object tracking - but I didn't spend a lot of time going through the docs.

**Reply by gabbymouth** [reply]
This is kind of a wiki for it with links:
`https://www.reddit.com/r/BlueIris/comments/1i9vcjb/official_blue_onyx_ai_for_blue_iris/` and the FAQ here:
`https://github.com/xnorpx/blue-onyx/discussions/14`

I saw this in the FAQ which may be of interest:
`Q: Why DETR models and not YOLO 8+?
A: Firstly, YOLO has an AGPL license from some sources, which creates a gray area for open source projects. I wanted to move away from that. Secondly, I aim to run state-of-the-art models, and DETR seems to be the future in this area. However Blue Onyx does support Yolo5 models.`

These are models that I can run with it:
`delivery.onnx
IPcam-animal.onnx
ipcam-bird.onnx
IPcam-combined.onnx
IPcam-dark.onnx
IPcam-general.onnx
package.onnx
rt-detrv2-l.onnx
rt-detrv2-m.onnx
rt-detrv2-ms.onnx
rt-detrv2-s.onnx
rt-detrv2-x.onnx`

But yeah, if object tracking is a requirement, then BO would not be a choice at this time.
